sdk/python/kfp/dsl/artifact.py

```diff
@@ -98,129 +46,105 @@ class Artifact(object):
+
   def __init__(self, instance_schema: Optional[str] = None):
-    """Constructs an instance of Artifact"""
+    """Constructs an instance of Artifact
+    Setups up self._metadata_fields to perform type checking and initialize RuntimeArtifact.
+    """
     if self.__class__ == Artifact:
       if not instance_schema:
         raise ValueError(
-            'The "instance_schema" argument must be set.')
-      schema_yaml = yaml.safe_load(instance_schema)
-      if 'properties' not in schema_yaml:
-        raise ValueError('Invalid instance_schema, properties must be present. '
-                         'Got %s' % instance_schema)
-      schema = schema_yaml['properties'] or {}
-      if 'title' not in schema_yaml:
-        raise ValueError('Invalid instance_schema, title must be present. '
-                         'Got %s' % instance_schema)
-      self.TYPE_NAME = schema_yaml['title']
-      self.PROPERTIES = {}
-      for k, v in schema.items():
-        self.PROPERTIES[k] = Property.from_dict(v)
+            'The "instance_schema" argument must be set for Artifact.')
+      self._instance_schema = instance_schema
     else:
       if instance_schema:
         raise ValueError(
-            'The "instance_schema" argument must not be passed for '
-            'Artifact subclass %s.' % self.__class__)
-      instance_schema = self.get_artifact_type()
+            'The "instance_schema" argument must not be passed for Artifact subclass: {}'.format(self.__class__))
+
+    # setup self._metadata_fields
+    self._parse_instance_schema()
 
-    # MLMD artifact type schema string.
-    self._type_schema = instance_schema
     # Instantiate a RuntimeArtifact pb message as the POD data structure.
     self._artifact = pipeline_spec_pb2.RuntimeArtifact()
+
+    # Stores the metadata for the Artifact.
+    self.metadata = {}
+
     self._artifact.type.CopyFrom(pipeline_spec_pb2.ArtifactTypeSchema(
-        instance_schema=instance_schema
+        instance_schema=self._instance_schema
     ))
-    # Initialization flag to prevent recursive getattr / setattr errors.
-    self._initialized = True
 
-  @classmethod
-  def get_artifact_type(cls) -> str:
-    """Gets the instance_schema according to the Python schema spec."""
-    title = _KFP_ARTIFACT_TITLE_PATTERN.format(cls.TYPE_NAME)
-    schema_map = None
-    if cls.PROPERTIES:
-      schema_map = {}
-      for k, v in cls.PROPERTIES.items():
-        schema_map[k] = {
-            'type': v.get_type_name(),
-            'description': v.description
-        }
-    result_map = {
-        'title': title,
-        'type': 'object',
-        'properties': schema_map
-    }
-    return serialization_utils.yaml_dump(result_map)
+    self._initialized = True
 
-  @classmethod
-  def get_ir_type(cls) -> pipeline_spec_pb2.ArtifactTypeSchema:
-    return pipeline_spec_pb2.ArtifactTypeSchema(
-        instance_schema=cls.get_artifact_type())
+  def set_custom_metadata_field(self, name: str, value: Any):
```

Comment:
@neuromage We had discussed to not have this method, however they are needed because: 

A user can do artifact.metadata['custom_property'] = 'test_value' and that needs to be reflected in the artifact.RuntimeArtifact.

This was discovered in testGetExecutorOutput in entrypoint_utils_test

Reply:
Actually figured a way to support what we decided and the entrypoint_util implementation.

Resolving this.